src/System.Private.CoreLib/src/System/Diagnostics/Eventing/EventPipeController.cs

```diff
@@ -0,0 +1,173 @@
+    /// if the trace file is /path/to/trace.netperf then the marker file is /path/to/trace.netperf.ctl.
+    /// This listener does not poll very often, and thus takes time to enable and disable tracing.  This is by design to ensure that the listener does
+    /// not starve other threads on the system.
+    /// NOTE: If COMPlus_EnableEventPipe != 4 then this listener is not created and does not add any overhead to the process.
+    /// </summary>
+    internal sealed class EventPipeController
+    {
+        private const string MarkerFileExtension = ".ctl";
+        private const int PollingIntervalMilliseconds = 10000; // 10 seconds
+        private const uint DefaultCircularBufferMB = 1024; // 1 GB
+        private readonly char[] ProviderConfigDelimiter = new char[] { ',' };
+        private readonly char[] ConfigComponentDelimiter = new char[] { ':' };
+
+        private readonly EventPipeProviderConfiguration[] DefaultProviderConfiguration = new EventPipeProviderConfiguration[]
+        {
+            new EventPipeProviderConfiguration("Microsoft-Windows-DotNETRuntime", 0x4c14fccbd, 5),
+            new EventPipeProviderConfiguration("Microsoft-Windows-DotNETRuntimePrivate", 0x4002000b, 5),
+            new EventPipeProviderConfiguration("Microsoft-DotNETCore-SampleProfiler", 0x0, 5)
+        };
+
+        private static EventPipeController s_controllerInstance;
+
+        private string m_traceFilePath;
+        private string m_markerFilePath;
+        private bool m_markerFileExists;
+
+        internal static void Initialize()
+        {
+            // Don't allow failures to propagate upstream.
+            // Instead, ensure program correctness without tracing.
+            try
+            {
+                if (s_controllerInstance == null)
+                {
+                    string strEnabledValue = CompatibilitySwitch.GetValueInternal("EnableEventPipe");
+                    if (strEnabledValue != null)
+                    {
+                        int enabledValue = Convert.ToInt32(strEnabledValue);
+                        if (enabledValue == 4)
+                        {
+                            s_controllerInstance = new EventPipeController();
+                        }
+                    }
+                }
+            }
+            catch { }
+        }
+
+        private EventPipeController()
+        {
+            // Determine the marker file path.
+            string traceFilePath = CompatibilitySwitch.GetValueInternal("EventPipeOutputFile");
+            if (!string.IsNullOrEmpty(traceFilePath))
+            {
+                m_traceFilePath = traceFilePath;
+                m_markerFilePath = traceFilePath + MarkerFileExtension;
+            }
+            else
+            {
+                // If the output file path is not specified then use
+                // path/to/current/working/directory/Process-<pid>.netperf.ctl
+                m_traceFilePath = EventPipeInternal.GetDefaultTraceFileName();
```

Comment:
There is a way to get the processID in managed code in system.private.corelib 
```
Win32Native.GetCurrentProcessId() 
```
EventSource already uses this.    You should use this and generate the file name here, rather than the pain of creating an FCALL to do it on the native side.

Reply:
Win32Native.GetCurrentProcessId() is a Windows-specific pinvoke:

```
        [DllImport(Interop.Libraries.Kernel32, CharSet = CharSet.Auto, SetLastError = true)]
        internal static extern uint GetCurrentProcessId();
```

If I remember correctly, this is used for some ETW-specific features.

If desired, we can expose the process ID via FCALL and do the work in managed code.

Reply:
I think we'll need to just plumb the PID to managed code to use it to resolve filename conflicts.

Reply:
Hmm, I saw that that definition was windows specific, but what I also know is that the call site for Win32Native.GetCurrentProcessId() is unconditional (EnsureDescriptorsInitialized has to be called even on Linux (or eventListeners don't work) and there are no ifdefs  around the call to GetCurrentProcessId.   

I was assuming on Linux we created look-alikes, but frankly I don't know how it works, so I could be wrong.   Still just try it and testing will show that it works on Linux (or you have to plumb through the FCALL).